### PR TITLE
fix: TimeZone issues on Wizard

### LINKED ISSUE #3582: Wrong Event Time on public page
The FOSSASIA page shows the event starts at 4pm. The event time in the wizard is: Start 9AM. I am using a browser in Germany.

Expected: Event time should show correct local time.
Screenshot from 2019-10-30 12-35-15

Wizard time of the same event:
Screenshot from 2019-10-30 12-38-25

### PR DESCRIPTION
**Pull request by kushthedude**
Fixes #3582

### CODE REVIEW

**Comment by kushthedude** [comment]
@prateekj117 @mariobehling @iamareebjamal Please check, Time was getting displayed wrongly on `Public Pages` due to helper issues but it was correct else where. Fixed it.

**Comment by iamareebjamal** [comment]
No, it is displaying correctly on public pages and wrong elsewhere

**Comment by kushthedude** [comment]
No I checked on my local. Every where else we are displaying raw time. But
in event page we are using the timezone to check the distance

**Comment by iamareebjamal** [comment]
2020-03-19T01:00:00+00:00" should be displayed as `March 19th 2020, 9:00 AM (+08)`

Which it is. So it is correct. Please correct the bug in Event Wizard which I have mentioned

**Comment by iamareebjamal** [comment]
We **DON'T** have to show raw time, we have to show time in event timezone. You have just created a PR to add a new bug which wasn't there instead of fixing the bug which is in the event wizard and dashboard

**Comment by kushthedude** [comment]
Hmmm, I got the time wrong, I thought the time is UTC and you have to
convert time to your time zone😌

**Comment by iamareebjamal** [comment]
The time is UTC, and you have to convert it to the event timezone

**Comment by kushthedude** [comment]
Got it.

**Comment by kushthedude** [comment]
> The time is UTC, and you have to convert it to the event timezone

I dont think so, the time which we enter in the Wizard asks the time in our timezone not the UTC. Therefore, whenever I enter time & date wizard as 9 Nov, 9:00 AM and select my timezone as Singapore. this mean its the time in 9:00 AM Singapore.

**Comment by kushthedude** [comment]
Also lets consider an example, I am organising an event in IST(+5:30) if my start time is 9:00AM , I will enter this time and select the timezone +5:30, This doesn't mean event wizard should make any changes to this time ?

**Comment by iamareebjamal** [comment]
> I dont think so, the time which we enter in the Wizard asks the time in our timezone not the UTC.

Did you take a look at the time returned by the server? Let's talk about facts and not air. Paste the time returned by the server

**Comment by iamareebjamal** [comment]
You're talking in abstractions, hypothetics and philosophies about timezones and I am telling you what is happening by giving you concrete examples and API responses and the link to event public page.

Let's continue this discussion with facts from this point on

**Comment by kushthedude** [comment]
> > I dont think so, the time which we enter in the Wizard asks the time in our timezone not the UTC.
> 
> Did you take a look at the time returned by the server? Let's talk about facts and not air. Paste the time returned by the server

I was talking in the air, I got your point and checked the server data. Event Public Page was showing the right time whereas others were showing wrong. Will update PR by tonight.

**Comment by kushthedude** [comment]
@iamareebjamal Please have a look, Fixed the time issues in events at every place of occuring.

**Comment by kushthedude** [comment]
Also standardised the way of displaying the time at every position with respect of events.

**Comment by iamareebjamal** [comment]
OK, please add before after screenshots and what about event wizard. I don't see any changes regarding that

File: app/helpers/general-date.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import Helper from '@ember/component/helper';
+import moment from 'moment';
+
+export function headerDate(params) {
+  let timezone = params[1] ? params[1] : moment.tz.guess();
```
Comment (iamareebjamal) [inline app/helpers/general-date.js]: Deconstruct params

**Comment by kushthedude** [comment]
> OK, please add before after screenshots and what about event wizard. I don't see any changes regarding that

Yes about the wizard, I tried adding 4-5 events with different Time zones, It was working fine only error was the one you higlighted of conversion of timezones.

File: app/helpers/general-date.js (hunk 1 above)
Comment (kushthedude) [reply app/helpers/general-date.js]: Shouldn't this rule only run on single allocation like
```js
let timezone = params[1]
``` 
in case of array

**Comment by iamareebjamal** [comment]
Yes, so where's that fixed?

**Comment by kushthedude** [comment]
> Yes, so where's that fixed?

No, I mean to say I didn't find error in wizard, error was in time displaying function which were not considering the timezones and were displaying raw time as fed into the db

**Comment by iamareebjamal** [comment]
OK, will confirm

**Comment by kushthedude** [comment]
@prateekj117 Please report if you find any other issue or bug related to the following.

**Comment by iamareebjamal** [comment]
No, it is still showing wrong time in event basic details wizard

**Comment by kushthedude** [comment]
> No, it is still showing wrong time in event basic details wizard

Can you please elaborate ?

**Comment by iamareebjamal** [comment]
Screenshot from 2019-11-10 20-55-44

Time in Event Dashboard is fixed. But format is wrong. Time format should be like this:

Screenshot from 2019-11-10 20-55-44

**Comment by iamareebjamal** [comment]
> Can you please elaborate ?

Screenshot from 2019-11-10 20-57-59

Should be 9 AM and 5 PM

**Comment by kushthedude** [comment]
> > Can you please elaborate ?
> 
> Screenshot from 2019-11-10 20-57-59
> 
> Should be 9 AM and 5 PM

When I am editing the event it should show 9 and 5 ?

**Comment by iamareebjamal** [comment]
Of course. It should say the same everywhere. In accordance with event timezone

**Comment by iamareebjamal** [comment]
@kushthedude Update?

**Comment by kushthedude** [comment]
Time being displayed o wizard is directly fed from db, Also we have kept
our wizard functionality not to complex, I am guessing introducing helpers
in them would be not a good option. Alternatively why dont we directly
store the final time in db after conversions of timezones . How does it
sound?

**Comment by iamareebjamal** [comment]
It sounds terrible. The original issue was simply that time displayed in event wizard is not same as in public pages. Are you saying that event wizard should show 9AM in Singapore Time and the value stored in database should be 4 PM in Singapore?

The major bug is this one. Every other bug is minor to extent of non-existence. I think you don't understand timezones. Either remove the event timezone dropdown or convert the time according to it and not the browser timezone

**Comment by kushthedude** [comment]
So, when I am creating a event time should be 9:00 and when I come to edit
it , it should show 11:30 . Is this was expected?

**Comment by kushthedude** [comment]
I will get back to this after I properly understand timezone issue and also
see how eventbrite handles the time.

**Comment by iamareebjamal** [comment]
> So, when I am creating a event time should be 9:00 and when I come to edit
it , it should show 11:30 . Is this was expected?

No, it should be 9:00 forever

**Comment by iamareebjamal** [comment]
@prateekj117 Can you please take up the event wizard part

**Comment by prateekj117** [comment]
@iamareebjamal On it.

**Comment by kushthedude** [comment]
@iamareebjamal I experimented with Eventbrite. Point to be noted from their :
- There is no timezone conversion or any such thing followed by them.
- While using Event Wizard, I just need to enter my time like if I entered `9:00 AM in Singapore Standard Time` time stored will be 9:00 AM and while displaying it will show `9:00 AM  Singapore Time`. 
- I also tried to see if there is interconversion of timezones so I used VPN's and tried to see that event, HOLA the time was getting shown `9:00 AM SGT` only.

Now, I don't understand why do we need to follow this interconversion. As I stated in my earlier points we should do the same and just change the time getting displayed on Public Page. 
Also if a user is entering 9:00 AM in our event wizard and selecting SGD as timezone, Why do we have to convert the time. Why don't we show the time as `9:00 AM in SGD` on every page?

Also, this interconversion will cause too much discrepancy in user, like currently if a user creates an event from India in SGD Timezone, Our app automatically adds 2:30 hours to the event time due to the timezone differences `8:00 and 5:30 UTC` which is not needed.

SS from the wizard and Public Event Page: 
Screenshot 2019-11-15 at 3 37 48 PM

Screenshot 2019-11-15 at 3 36 55 PM

**Comment by iamareebjamal** [comment]
> I just need to enter my time like if I entered 9:00 AM in Singapore Standard Time time stored will be 9:00 AM and while displaying it will show 9:00 AM Singapore Time

That's what we want. Show me that it is done in eventyay. 9AM should be shown as 9AM everywhere. Is this happening right now? No. So what should be done?

Right now, 9AM in Singapore is shown as something else in Public Pages. So you just proved my point. Once, you have achieved this behaviour, we can close the issue. Thanks

**Comment by iamareebjamal** [comment]
> currently if a user creates an event from India in SGD Timezone, Our app automatically adds 2:30 hours to the event time due to the timezone differences 8:00 and 5:30 UTC which is not needed.

Exactly, thank you. It should not add 2:30 or anything, it should send the time with event timezone, regardless of the timezone of the browser. 

You're just repeating what I said previously. We don't want to send 9AM +5:30 which we are currently, we want to send 9AM +8:00 (Singapore Time) 

So, I don't know what you're trying to say what I already am not saying

**Comment by kushthedude** [comment]
> > I just need to enter my time like if I entered 9:00 AM in Singapore Standard Time time stored will be 9:00 AM and while displaying it will show 9:00 AM Singapore Time
> 
> That's what we want. Show me that it is done in eventyay. 9AM should be shown as 9AM everywhere. Is this happening right now? No. So what should be done?
> 
> Right now, 9AM in Singapore is shown as something else in Public Pages. So you just proved my point. Once, you have achieved this behaviour, we can close the issue. Thanks

In my first commit , I made this change only to remove the timezone interconversion and ensure the time entered in wizard is displahed as it is. 
Since we are now clear that we need only this type of time displaying, I will fix it up

**Comment by iamareebjamal** [comment]
No, your changes added a new issue than fixing it. You don't understand timezones. If your changes show correct time everywhere, then continue. I guarantee you they'll not. Set 9 AM SGT in event wizard and show me the time everywhere. I'll wait

**Comment by kushthedude** [comment]
Ohk, Now I understand the conplete scenario, Start time getting stored in
DB contains the timezone of the browser and hence db stores the incorrect
time as you stated. But when we are displaying the time on FE we ignore the
timezone part of start time and use timezone given by user.
But the db should not store the incorrect timezone in the first place with
start time.
I will fix this and now I understand the error you were referring to in
wizard.
@iamareebjamal

**Comment by iamareebjamal** [comment]
> But when we are displaying the time on FE we ignore the
timezone part of start time and use timezone given by user.

We don't ignore the timezone part. @prateekj117 has made some progress on this. So, wait for him to report how much he has done

**Comment by prateekj117** [comment]
@iamareebjamal I am working on correctly displaying the time on frontend now. The time sent to the server is now being sent in the event timezone. Will try my best to complete it by tonight.

**Comment by prateekj117** [comment]
@iamareebjamal Should I also make a new component, or should I just convert the timezone when rendered.

**Comment by iamareebjamal** [comment]
When rendered

**Comment by iamareebjamal** [comment]
@kushthedude Please fix the time format so we can merge this. Another PR is open for event details wizard

**Comment by kushthedude** [comment]
Sure

**Comment by iamareebjamal** [comment]
@kushthedude Updates? This is blocking the release now. You stay active each day and ask assignment on issues but rarely update on PRs

**Comment by kushthedude** [comment]
I was waiting on merge of other the PR, I will update this today.

File: app/helpers/general-date.js (hunk 1 above)
Comment (kushthedude) [reply app/helpers/general-date.js]: @iamareebjamal Please update me on this rule, I couldn't find what sort of change I need to make for deconstruct params

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- tests/integration/helpers/general-date-test.js  1
- app/helpers/general-date.js  1
         

Complexity decreasing per file
==============================
+ app/helpers/header-date.js  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/open-event-frontend/open-event-frontend/pullRequest?prid=4463702&bid=15077688)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/helpers/general-date.js  1
- tests/integration/helpers/general-date-test.js  1
         

Complexity decreasing per file
==============================
+ app/helpers/header-date.js  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/open-event-frontend/open-event-frontend/pullRequest?prid=4463702&bid=15077688)

**Comment by kushthedude** [comment]
@iamareebjamal This PR is delayed, But please give me a little more time to check if everything is working as expected or not. As by my approval the recent PR broke the complete event dashboard.

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/helpers/general-date.js  2
- tests/integration/helpers/general-date-test.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/open-event-frontend/open-event-frontend/pullRequest?prid=4463702&bid=15077688)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/helpers/general-date.js  2
- tests/integration/helpers/general-date-test.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/open-event-frontend/open-event-frontend/pullRequest?prid=4463702&bid=15077688)

**Comment by kushthedude** [comment]
@iamareebjamal Please review  🎉

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/helpers/general-date.js  2
- tests/integration/helpers/general-date-test.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/open-event-frontend/open-event-frontend/pullRequest?prid=4463702&bid=15077688)

**Comment by iamareebjamal** [comment]
https://github.com/fossasia/open-event-frontend/pull/3611#issuecomment-552204584

**Comment by kushthedude** [comment]
> [#3611 (comment)](https://github.com/fossasia/open-event-frontend/pull/3611#issuecomment-552204584)

I have followed the same format, just added the timezone in suffix for ease of time.

**Comment by iamareebjamal** [comment]
Some other formats have been removed or changed? Anyway, I'll test after coming from office

**Comment by kushthedude** [comment]
> Some other formats have been removed or changed? 

Except at public event page, at every other place we had different format for event time as dicussed on chat I have used a single helper to generalise date format for events.



> Anyway, I'll test after coming from office

Sure let me know if any more thing needs to be changed 😄

**Comment by iamareebjamal** [comment]
> we had different format for event time 

In some places, time was displayed, in some places, date. So, different formats should be used for both. But not more than 2

**Comment by kushthedude** [comment]
> > we had different format for event time
> 
> In some places, time was displayed, in some places, date. So, different formats should be used for both. But not more than 2

No, Please see the changes at every place where StartTime was getting shown it was shown with date and time. Event the `Cell General Date` which is responsible for rendering time at most places was not consistent, As we were bypassing the format by passing `DateFormat` from ember table. All I have done is made the Time and Date consistent to format `HH:MM, Date (TimeZone)`

**Comment by iamareebjamal** [comment]
> Local (time correct, format wrong)
> 
> Screenshot from 2019-11-10 20-55-44
> 
> Time in Event Dashboard is fixed. But format is wrong. Time format should be like this:
> 
> Production (time wrong, format correct)
> Screenshot from 2019-11-10 20-57-10

See the difference here

**Comment by iamareebjamal** [comment]
We don't need Thursday everywhere

**Comment by kushthedude** [comment]
I have not shown day everywhere 😅

**Comment by iamareebjamal** [comment]
Great work!

**Comment by kushthedude** [comment]
> Great work!

Time for a release 🤘 @iamareebjamal

**Comment by iamareebjamal** [comment]
Not till debounce in search is fixed

**Comment by kushthedude** [comment]
> Not till debounce in search is fixed

I will take a look at debounce

**Pull request merged**